### PR TITLE
oobmigrations: Run in `worker`, not `frontend`

### PR DESCRIPTION
**Pull request by efritz**
There's no reason to do backround work in our API servers, especially since this is work that puts some consistent load on the database when running. Let's move it to a place where the horizontal scalability is controlled and not directly controlled by concurrent user count.

Environment variables that move from `frontend` to `worker`:

- [`ALLOW_DECRYPT_MIGRATION`](https://k8s.sgdev.org/search?q=context:global+ALLOW_DECRYPT_MIGRATION&patternType=literal)
- [`SRC_DISABLE_OOBMIGRATION_VALIDATION`](https://k8s.sgdev.org/search?q=context:global+SRC_DISABLE_OOBMIGRATION_VALIDATION&patternType=literal)
- [`PRECISE_CODE_INTEL_.*_MIGRATION_.*`](https://k8s.sgdev.org/search?q=context:global+PRECISE_CODE_INTEL_.*_MIGRATION_.*&patternType=regexp)

### CODE REVIEW

File: cmd/worker/shared/migrations.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+	"github.com/cockroachdb/errors"
+
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/oobmigration"
+)
+
+func registerOSSMigrations(db database.DB, outOfBandMigrationRunner *oobmigration.Runner) error {
+	// Run a background job to handle encryption of external service configuration.
+	extsvcMigrator := oobmigration.NewExternalServiceConfigMigratorWithDB(db)
```
Comment (eseliger) [inline cmd/worker/shared/migrations.go:15]: I think I would move these migrator codes into the worker codebase as well.

Comment (efritz) [reply cmd/worker/shared/migrations.go:15]: Updated.

File: cmd/worker/internal/migrations/init.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	outOfBandMigrationRunner := oobmigration.NewRunnerWithDB(db, time.Second*30, observationContext)
+
+	if err := m.registerMigrators(db, outOfBandMigrationRunner); err != nil {
+		return nil, err
+	}
+
+	if os.Getenv("SRC_DISABLE_OOBMIGRATION_VALIDATION") != "" {
+		log15.Warn("Skipping out-of-band migrations check")
+	} else {
+		// See #29074
```
Comment (efritz) [inline cmd/worker/internal/migrations/init.go:64]: #29074

**Review by coury-clark — APPROVED** [review]

**Comment by coury-clark** [comment]
Awesome to see this! Thanks! ❤️

File: cmd/worker/shared/main.go [hunk 3]
```diff
@@ -246,3 +258,17 @@ func jobNames(jobs map[string]job.Job) []string {
 
 	return names
 }
+
+func composeRegisterMigrations(fns ...func(db database.DB, outOfBandMigrationRunner *oobmigration.Runner) error) func(db database.DB, outOfBandMigrationRunner *oobmigration.Runner) error {
+	return func(db database.DB, outOfBandMigrationRunner *oobmigration.Runner) error {
+		for _, fn := range fns {
+			if fn != nil {
+				if err := fn(db, outOfBandMigrationRunner); err != nil {
+					return err
```
[the remark this answers is not in the document]
Comment (efritz) [reply cmd/worker/shared/main.go:267]: Yeah these should end up being fatal errors.

File: doc/admin/how-to/unfinished_migration.md [hunk 4]
```diff
@@ -28,7 +28,7 @@ ERROR: Unfinished migrations. Please revert Sourcegraph to the previous version
 
 Each migration notes the engineering team that is able to resolve associated errors.
 
-As an escape hatch, the environment variable `SRC_DISABLE_OOBMIGRATION_VALIDATION` can be set to `true` to disable the startup check.
+As an escape hatch, the environment variable `SRC_DISABLE_OOBMIGRATION_VALIDATION` can be set to `true` on the `frontend` and `worker` services to disable the startup check.
```
[the remark this answers is not in the document]
Comment (efritz) [reply doc/admin/how-to/unfinished_migration.md:31]: No, but #29074 (linked in a comment above) might extend some functionality back to the frontend on startup, and saying we require it on both just covers our bases from missing this change again.

**Review by eseliger — APPROVED** [review]
Nice work 🚀 Thanks for doing this!!

**Review by caugustus-sourcegraph — APPROVED** [review]
Yay! It doesn't look like any of the env var's mentioned here are used in any of our deployment repos, so should be good there.

**Pull request merged**